### PR TITLE
Add XPath library for parsing xpath config paths

### PR DESCRIPTION
**Pull request by wingo**
From the original #577:

> This library adds an library which parses a small subset of xpath queries against a yang schema. The library currently supports querying arrays (specified as `position()=1` for the first element) based on the position, querying lists based on keys (specified as `[addr=1.2.3.4]`) and specifying leaves and containers based on their name (more information in documentation at the top of the file).
> 
> The path is transformed into a lua path (as suggested in #549). The only difference is that instead of just being a table of keys or just being an integer it's a table with a "name" field and a key or keys field. This change is because our query is XPath and not the adhoc query language which specified paths sligthly differently.

This PR is against `lwaftr` instead of `yang`.

### CODE REVIEW

File: src/apps/config/leader.lua [hunk 1]
```diff
@@ -93,22 +94,86 @@ function Leader:rpc_describe (args)
    return { native_schema = self.schema_name }
 end
 
+local function path_getter_for_grammar(grammar, path)
+   return path_resolver(grammar, path)
+end
```
Comment (tsyesika) [inline src/apps/config/leader.lua]: Not sure what the point in this function is, it just calls path_resolver without doing anything.

File: src/lib/yang/path.lua [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+--    position in the array.
+-- Table fields:
+--    This is a table which has a "name" property specifying member
+--    name and has a "keys" (not key) property which is either:
+--       - A string representing the key if the table is string keyed.
+--       - A lua table with corrisponding leaf names as the key and the
+--         value as the value.
+module(..., package.seeall)
+
+local ffi = require("ffi")
```
Comment (tsyesika) [inline src/lib/yang/path.lua]: This isn't used anymore

**Review by tsyesika — COMMENTED** [review]
Everything looks good, one function doesn't seem to do anything which i commented on and the ffi library looks like it's unneeded now but otherwise it all looks and seems to work well :)

File: src/apps/config/leader.lua (hunk 1 above)
Comment (wingo) [reply src/apps/config/leader.lua]: indeed, it was a placeholder that just got filled in when the path support landed. Fixed.

**Pull request merged**